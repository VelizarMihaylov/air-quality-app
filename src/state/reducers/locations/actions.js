export const REMOVE_LOCATION = 'REMOVE_LOCATION'

export const removeLocationAction = location => ({
  type: REMOVE_LOCATION,
  payload: location
})
