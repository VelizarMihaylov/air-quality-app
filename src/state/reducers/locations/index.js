import { REMOVE_LOCATION } from './actions'

const locations = (store = [], action) => {
  switch (action.type) {
    case 'ADD_LOCATIONS':
      return action.payload
    case REMOVE_LOCATION:
      return [...store.filter(({ location }) => location !== action.payload)]
    default:
      return store
  }
}

export default locations
