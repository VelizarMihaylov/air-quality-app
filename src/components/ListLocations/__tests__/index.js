import React from 'react'
import { ListLocations } from '../index'
import { create, act } from 'react-test-renderer'
import { Provider } from 'react-redux'

import state from 'state'
import locations from '../__mocks__'

import dayjs from 'dayjs'
jest.mock('dayjs')

describe('ListLocations', () => {
  dayjs.mockImplementation(() => ({
    extend: jest.fn(),
    utc: () => ({
      fromNow: () => '5 hours ago'
    })
  }))

  it('should handle loading state', () => {
    const ListLocationsRender = create(<ListLocations loading dispatch={jest.fn()} locations={[]} />).toJSON()
    expect(ListLocationsRender).toMatchSnapshot()
  })
  it('should handle error state', () => {
    const ListLocationsRender = create(<ListLocations error dispatch={jest.fn()} locations={[]} />).toJSON()
    expect(ListLocationsRender).toMatchSnapshot()
  })
  it('should render locations when not in loading or error state and locations prop has length', () => {
    const ListLocationsRender = create(
      <Provider store={state}>
        <ListLocations
          dispatch={jest.fn()}
          data={null}
          city='Manchester'
          locations={locations}
        />
      </Provider>
    ).toJSON()
    expect(ListLocationsRender).toMatchSnapshot()
  })
  it('should render locations when not in loading or error state and locations prop has length', () => {
    const ListLocationsRender = create(
      <Provider store={state}>
        <ListLocations
          dispatch={jest.fn()}
          data={null}
          city='Manchester'
          locations={locations}
        />
      </Provider>
    ).toJSON()
    expect(ListLocationsRender).toMatchSnapshot()
  })
  it('should dispatch ADD_LOCATIONS action with the locations received as data when the city props is updated', () => {
    const ListLocationsRender = create(
      <Provider store={state}>
        <ListLocations
          key='test'
          dispatch={jest.fn()}
          data={null}
          city={null}
          locations={[]}
        />
      </Provider>
    )

    act(() => ListLocationsRender.update(
      <Provider store={state}>
        <ListLocations
          key='test'
          dispatch={jest.fn()}
          data={{ results: locations }}
          city='Manchester'
          locations={[]}
        />
      </Provider>
    ))
    expect(ListLocationsRender.root.props.children.props.dispatch).toHaveBeenCalledWith({
      type: 'ADD_LOCATIONS',
      payload: locations
    })
  })
})
