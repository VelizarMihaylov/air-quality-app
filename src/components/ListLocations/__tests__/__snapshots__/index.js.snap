// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListLocations should handle error state 1`] = `
<h1>
  Oops something went wrong!
</h1>
`;

exports[`ListLocations should handle loading state 1`] = `
<div
  className="ListLocations"
>
  <div
    className="ListLocations--spinner"
  />
</div>
`;

exports[`ListLocations should render locations when not in loading or error state and locations prop has length 1`] = `
<div
  className="ListLocations"
>
  <div
    className="LocationCard"
  >
    <svg
      className="LocationCard--icon__close"
      id="Capa_1"
      onClick={[Function]}
      style={
        Object {
          "enableBackground": "new 0 0 348.333 348.334",
        }
      }
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xlink="http://www.w3.org/1999/xlink"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="LocationCard--content"
    >
      <p
        className="LocationCard--content--updatedAt"
      >
        Updated 50 years ago
      </p>
      <h3>
        Manchester Piccadilly
      </h3>
      <p>
        in Manchester, United Kingdom
      </p>
      <p
        className="LocationCard--content--values"
      >
        Value: NO2: 37, SO2: 3, PM25: 4, O3: 9
      </p>
    </div>
  </div>
  <div
    className="LocationCard"
  >
    <svg
      className="LocationCard--icon__close"
      id="Capa_1"
      onClick={[Function]}
      style={
        Object {
          "enableBackground": "new 0 0 348.333 348.334",
        }
      }
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xlink="http://www.w3.org/1999/xlink"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="LocationCard--content"
    >
      <p
        className="LocationCard--content--updatedAt"
      >
        Updated 50 years ago
      </p>
      <h3>
        Manchester Sharston
      </h3>
      <p>
        in Manchester, United Kingdom
      </p>
      <p
        className="LocationCard--content--values"
      >
        Value: O3: 18, NO2: 16
      </p>
    </div>
  </div>
  <div
    className="LocationCard"
  >
    <svg
      className="LocationCard--icon__close"
      id="Capa_1"
      onClick={[Function]}
      style={
        Object {
          "enableBackground": "new 0 0 348.333 348.334",
        }
      }
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xlink="http://www.w3.org/1999/xlink"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="LocationCard--content"
    >
      <p
        className="LocationCard--content--updatedAt"
      >
        Updated 4 hours ago
      </p>
      <h3>
        Salford Eccles
      </h3>
      <p>
        in Manchester, United Kingdom
      </p>
      <p
        className="LocationCard--content--values"
      >
        Value: NO2: 30, PM25: 7, PM10: 12
      </p>
    </div>
  </div>
</div>
`;

exports[`ListLocations should render locations when not in loading or error state and locations prop has length 2`] = `
<div
  className="ListLocations"
>
  <div
    className="LocationCard"
  >
    <svg
      className="LocationCard--icon__close"
      id="Capa_1"
      onClick={[Function]}
      style={
        Object {
          "enableBackground": "new 0 0 348.333 348.334",
        }
      }
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xlink="http://www.w3.org/1999/xlink"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="LocationCard--content"
    >
      <p
        className="LocationCard--content--updatedAt"
      >
        Updated 50 years ago
      </p>
      <h3>
        Manchester Piccadilly
      </h3>
      <p>
        in Manchester, United Kingdom
      </p>
      <p
        className="LocationCard--content--values"
      >
        Value: NO2: 37, SO2: 3, PM25: 4, O3: 9
      </p>
    </div>
  </div>
  <div
    className="LocationCard"
  >
    <svg
      className="LocationCard--icon__close"
      id="Capa_1"
      onClick={[Function]}
      style={
        Object {
          "enableBackground": "new 0 0 348.333 348.334",
        }
      }
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xlink="http://www.w3.org/1999/xlink"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="LocationCard--content"
    >
      <p
        className="LocationCard--content--updatedAt"
      >
        Updated 50 years ago
      </p>
      <h3>
        Manchester Sharston
      </h3>
      <p>
        in Manchester, United Kingdom
      </p>
      <p
        className="LocationCard--content--values"
      >
        Value: O3: 18, NO2: 16
      </p>
    </div>
  </div>
  <div
    className="LocationCard"
  >
    <svg
      className="LocationCard--icon__close"
      id="Capa_1"
      onClick={[Function]}
      style={
        Object {
          "enableBackground": "new 0 0 348.333 348.334",
        }
      }
      version="1.1"
      viewBox="0 0 348.333 348.334"
      x="0px"
      xlink="http://www.w3.org/1999/xlink"
      xmlSpace="preserve"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M336.559,68.611L231.016,174.165l105.543,105.549c15.699,15.705,15.699,41.145,0,56.85c-7.844,7.844-18.128,11.769-28.407,11.769c-10.296,0-20.581-3.919-28.419-11.769L174.167,231.003L68.609,336.563c-7.843,7.844-18.128,11.769-28.416,11.769c-10.285,0-20.563-3.919-28.413-11.769c-15.699-15.698-15.699-41.139,0-56.85l105.54-105.549L11.774,68.611c-15.699-15.699-15.699-41.145,0-56.844c15.696-15.687,41.127-15.687,56.829,0l105.563,105.554L279.721,11.767c15.705-15.687,41.139-15.687,56.832,0C352.258,27.466,352.258,52.912,336.559,68.611z"
        />
      </g>
    </svg>
    <div
      className="LocationCard--content"
    >
      <p
        className="LocationCard--content--updatedAt"
      >
        Updated 4 hours ago
      </p>
      <h3>
        Salford Eccles
      </h3>
      <p>
        in Manchester, United Kingdom
      </p>
      <p
        className="LocationCard--content--values"
      >
        Value: NO2: 30, PM25: 7, PM10: 12
      </p>
    </div>
  </div>
</div>
`;
